File: 5_NthEnd/solution.c
#include <stdio.h>
#include <stdlib.h>

struct Node {
    int data;
    struct Node* next;
};

int getNthFromLast(struct Node* head, int n) {
    struct Node* first = head;
    struct Node* second = head;
    for (int i = 0; i < n; i++) {
        if (!first) return -1;
        first = first->next;
    }
    while (first) {
        first = first->next;
        second = second->next;
    }
    return second ? second->data : -1;
}


int main() {
    struct Node* head = NULL;
    int n, val, pos;
    printf("Enter number of elements: ");
    scanf("%d", &n);
    for (int i = 0; i < n; i++) {
        printf("Enter value: ");
        scanf("%d", &val);
        head = insertEnd(head, val);
    }

    printf("Enter N (from end): ");
    scanf("%d", &pos);
    int result = getNthFromLast(head, pos);
    printf("Nth node from end: %d\n", result);

    printList(head);
    return 0;
}

File: 5_NthEnd/solution.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

def getNthFromLast(head, n):
    first = head
    second = head
    for _ in range(n):
        if not first:
            return -1
        first = first.next
    while first:
        first = first.next
        second = second.next
    return second.data if second else -1


def build_list():
    head = None
    tail = None
    n = int(input("Enter number of elements: "))
    for _ in range(n):
        val = int(input("Enter value: "))
        new_node = Node(val)
        if not head:
            head = tail = new_node
        else:
            tail.next = new_node
            tail = new_node
    return head

def main():
    head = build_list()

    n = int(input("Enter N (from end): "))
    print("Nth node from end:", getNthFromLast(head, n))

    printList(head)

if __name__ == "__main__":
    main()